pallets/vector/src/lib.rs

```diff
@@ -780,6 +816,130 @@ pub mod pallet {
 			Self::deposit_event(Event::<T>::NewUpdater { old, new: updater });
 			Ok(())
 		}
+
+		/// The entrypoint for fulfill a call.
+		/// proof Input proof value.
+		/// public_values Input public values.
+		#[pallet::call_index(13)]
+		#[pallet::weight(T::WeightInfo::fulfill())]
+		pub fn fulfill(
+			origin: OriginFor<T>,
+			proof: ProofInput,
+			public_values: PublicValuesInput,
+		) -> DispatchResultWithPostInfo {
+			let sender: [u8; 32] = ensure_signed(origin)?.into();
+			let updater = Updater::<T>::get();
+			// ensure sender is preconfigured
+			ensure!(H256(sender) == updater, Error::<T>::UpdaterMisMatch);
+
+			let proof_outputs: ProofOutputs = SolValue::abi_decode(&public_values, true)
+				.map_err(|_| Error::<T>::CannotDecodePublicValue)?;
+
+			let head = Head::<T>::get();
+			let new_head: u64 = proof_outputs.newHead.to();
+			ensure!(new_head > head, Error::<T>::SlotBehindHead);
+
+			let sp1_vk = SP1VerificationKey::<T>::get();
+
+			let is_valid = Groth16Verifier::verify(
+				&proof,
+				&public_values,
+				&format!("{:?}", sp1_vk),
+				&GROTH16_VK_BYTES,
+			);
+			ensure!(is_valid.is_ok(), Error::<T>::VerificationFailed);
```

Comment:
Note: Would add a comment here about how this leads to unexpected behavior due to the `unwrap()`'s in v3, but will be resolved in v4.